### PR TITLE
Profile Subcommands

### PR DESCRIPTION
**Pull request by jamestouri**
https://github.com/medplum/medplum/issues/2469

### CODE REVIEW

**Comment by reshmakh** [comment]
Perhaps add some quick docs here on how to use?  Need not be comprehensive, I can touch it up - https://www.medplum.com/docs/cli/external-fhir-servers

**Comment by reshmakh** [comment]
@jamestouri - got it

File: packages/cli/src/profiles.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+removeProfile
+  .argument('<profileName>', 'Name of the profile')
+  .description('Remove a profile by name')
+  .action(async (profileName) => {
+    const storage = new FileSystemStorage(profileName);
+    storage.setObject('profile', undefined);
+    console.log(`${profileName} profile removed`);
+  });
+
+listProfiles.description('All profiles saved').action(async () => {
```
Comment (codyebberson) [inline packages/cli/src/profiles.ts:37]: Description should probably start with a verb, "Lists all profiles saved"

File: packages/cli/src/profiles.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    if (profile) {
+      allProfiles.push({ profileName: fileName, profile });
+    }
+  });
+  console.log(allProfiles);
+});
+
+describeProfile
+  .argument('<profileName>', 'Name of the profile')
+  .description('Properties of a profile')
```
Comment (codyebberson) [inline packages/cli/src/profiles.ts:54]: Description should probably start with a verb, "Describes a profile"

File: packages/cli/src/util/command.ts [hunk 3]
```diff
@@ -7,5 +7,6 @@ export function createMedplumCommand(name: string): Command {
     .option('--base-url <baseUrl>', 'FHIR server base url')
     .option('--token-url <tokenUrl>', 'FHIR server token url')
     .option('--authorize-url <authorizeUrl>', 'FHIR server authorize url')
-    .option('--fhir-url-path <fhirUrlPath>', 'FHIR server url path');
+    .option('--fhir-url-path <fhirUrlPath>', 'FHIR server url path')
+    .option('--auth-type <authType>', 'Type of authentication', 'basic');
```
Comment (codyebberson) [inline packages/cli/src/util/command.ts:11]: Is this where we can list all auth types?  We should determine our list, and the exact strings we want to accept?

We have this enum, which might be a good starting point:

```ts
export enum OAuthGrantType {
  ClientCredentials = 'client_credentials',
  AuthorizationCode = 'authorization_code',
  RefreshToken = 'refresh_token',
  JwtBearer = 'urn:ietf:params:oauth:grant-type:jwt-bearer',
  TokenExchange = 'urn:ietf:params:oauth:grant-type:token-exchange',
}
```

So maybe:

- `basic` - Basic auth
- `client_credentials` - Client credentials flow
- `authorization_code` - Authorization code flow (default)
- `jwt-bearer` - User provided presigned JWT
- `token-exchange` - Preconfigured external auth provider

File: packages/cli/src/profiles.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  .addCommand(removeProfile)
+  .addCommand(listProfiles)
+  .addCommand(describeProfile);
+
+setProfile
+  .argument('<profileName>', 'Name of the profile')
+  .description('Create a new profile or replace it with the given name and its associated properties')
+  .action(async (profileName, options) => {
+    const storage = new FileSystemStorage(profileName);
+    storage.setObject('profile', options);
```
Comment (codyebberson) [inline packages/cli/src/profiles.ts:24]: Thoughts on whether this property should be called "profile" vs "options" vs "config" vs ???

Comment (jamestouri) [reply packages/cli/src/profiles.ts:24]: setting to "options"

**Review by codyebberson — APPROVED** [review]

**Pull request merged**